### PR TITLE
fix(influxdb): check if fields are empty before sending

### PR DESCRIPTION
**Pull request by thalesmg**
Related Issue: https://emqx.atlassian.net/browse/EMQX-8461

Currently, the InfluxDB client raises an error if an empty `fields` map is passed to it for pushing data.

```
  14:03:35.563 [error] [InfluxDB] Encode [
    %{
      fields: %{},
      measurement: "t/topic",
      tags: %{},
      timestamp: 1670864615563
    }
  ] failed: :error :missing_field [
    {:influxdb_line, :encode_fields, 1,
     [
       file: '/home/thales/dev/emqx/emqx/deps/influxdb/src/influxdb_line.erl',
       line: 60
     ]},
    {:influxdb_line, :encode_, 1,
     [
       file: '/home/thales/dev/emqx/emqx/deps/influxdb/src/influxdb_line.erl',
       line: 44
     ]},
    {:influxdb_line, :"-encode_/1-fun-0-", 2,
     [
       file: '/home/thales/dev/emqx/emqx/deps/influxdb/src/influxdb_line.erl',
       line: 38
     ]},
    {:influxdb, :write, 2,
     [file: '/home/thales/dev/emqx/emqx/deps/influxdb/src/influxdb.erl', line: 79]},
    {:emqx_ee_connector_influxdb, :do_query, 3,
     [
       file: '/home/thales/dev/emqx/emqx/lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl',
       line: 322
     ]},
    {:emqx_resource_worker, :apply_query_fun, 6,
     [
       file: '/home/thales/dev/emqx/emqx/apps/emqx_resource/src/emqx_resource_worker.erl',
       line: 454
     ]},
    {:emqx_resource_worker, :query_or_acc, 3,
     [
       file: '/home/thales/dev/emqx/emqx/apps/emqx_resource/src/emqx_resource_worker.erl',
       line: 306
     ]},
    {:gen_statem, :loop_state_callback, 11, [file: 'gen_statem.erl', line: 1203]},
    {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 226]}
  ]
  2022-12-12T14:03:35.563607-03:00 [error] [InfluxDB] Encode [#{fields => #{},measurement => <<"t/topic">>,tags => #{},timestamp => 1670864615563}] failed: error missing_field [{influxdb_line,encode_fields,1,[{file,"/home/thales/dev/emqx/emqx/deps/influxdb/src/influxdb_line.erl"},{line,60}]},{influxdb_line,encode_,1,[{file,"/home/thales/dev/emqx/emqx/deps/influxdb/src/influxdb_line.erl"},{line,44}]},{influxdb_line,'-encode_/1-fun-0-',2,[{file,"/home/thales/dev/emqx/emqx/deps/influxdb/src/influxdb_line.erl"},{line,38}]},{influxdb,write,2,[{file,"/home/thales/dev/emqx/emqx/deps/influxdb/src/influxdb.erl"},{line,79}]},{emqx_ee_connector_influxdb,do_query,3,[{file,"/home/thales/dev/emqx/emqx/lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl"},{line,322}]},{emqx_resource_worker,apply_query_fun,6,[{file,"/home/thales/dev/emqx/emqx/apps/emqx_resource/src/emqx_resource_worker.erl"},{line,454}]},{emqx_resource_worker,query_or_acc,3,[{file,"/home/thales/dev/emqx/emqx/apps/emqx_resource/src/emqx_resource_worker.erl"},{line,306}]},{gen_statem,loop_state_callback,11,[{file,"gen_statem.erl"},{line,1203}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]

  14:03:35.565 [error] [connector: "bridge:influxdb_api_v2:aaa", msg: 'influxdb write point failed', reason: :missing_field]
  2022-12-12T14:03:35.565345-03:00 [error] connector: <<"bridge:influxdb_api_v2:aaa">>, line: 335, mfa: emqx_ee_connector_influxdb:do_query/3, msg: influxdb write point failed, reason: missing_field
  []

  14:03:35.565 [error] [id: "bridge:influxdb_api_v2:aaa", msg: :send_error, reason: :missing_field]
  iex(emqx@127.0.0.1)2> 2022-12-12T14:03:35.565913-03:00 [error] id: <<"bridge:influxdb_api_v2:aaa">>, line: 396, mfa: emqx_resource_worker:handle_query_result/4, msg: send_error, reason: missing_field
```

Instead of raising, we check if the interpolation resulted in an empty map due to lack of context and just fail the push more gracefully.

Related to this, the original issue _appears_ to be related to a frontend issue (to be confirmed and fixed separately), where the it is not encoding the field types:
https://emqx.atlassian.net/browse/EMQX-8461?focusedCommentId=24805

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/` dir
- [ ] For EMQX 4.x: `appup` files updated (execute `scripts/update-appup.sh emqx`)
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up
- [ ] In case of non-backward compatible changes, reviewer should check this item as a write-off, and add details in **Backward Compatibility** section

### CODE REVIEW

**Comment by JimMoen** [comment]
maybe a function case is better.

```erlang
lines_to_points(Data, [#{fields := Fields} | Rest], ResultPointsAcc, ErrorPointsAcc)
  when map_size(Fields) =:= 0 ->
    lines_to_points(Data, Rest, ResultPointsAcc, [{error, no_fields} | ErrorPointsAcc]);
```

File: lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl [hunk 1]
```diff
@@ -446,15 +446,23 @@ lines_to_points(
...
-                fields => EncodeFields
+                tags => EncodedTags,
+                fields => EncodedFields
             },
-            lines_to_points(Data, Rest, [Point | ResultPointsAcc], ErrorPointsAcc);
+            case map_size(EncodedFields) =:= 0 of
+                true ->
+                    %% influxdb client doesn't like empty field maps...
+                    lines_to_points(Data, Rest, ResultPointsAcc, [
+                        {error, no_fields} | ErrorPointsAcc
```
Comment (zmstone) [inline lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:461]: Not the scope of this pr
If there are encoding errors, should we really discard all?
Maybe just log it and move on?

Comment (JimMoen) [reply lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:461]: It's for `batch_query`
Partial success can lead to inaccurate resource metrics when batching query.

**Comment by thalesmg** [comment]
> maybe a function case is better.
> 
> ```erlang
> lines_to_points(Data, Lines, [#{fields := Fields} | Rest] = _ResultPointsAcc, ErrorPointsAcc)
>   when map_size(Fields) =:= 0 ->
>     lines_to_points(Data, Lines, Rest, [{error, no_fields} | ErrorPointsAcc]);
> ```

In my opinion, checking if the result of the _previous_ iteration worked in the _next_ iteration and changing the accumulator then is more confusing than doing it in the iteration which does the transformation.  :thinking:

**Review by zmstone — APPROVED** [review]

**Review by JimMoen — APPROVED** [review]

**Pull request merged**